FarihaIS/VanRoomies
# enhancement(auth): Token Authentication and OAuth Google Verification

# Welcome to VanRoomies! 👋✈️

Fixes: #32 

## Description of the change:
This change is an enhancement on our backend service for Google Sign In taking place at the front end. It enables the backend server the following:
- Authenticate/verify that the Google Sign In request originates from a Google account user.
- Authenticate and authorize subsequent requests to user-private resources via a [Json Web Token](https://www.npmjs.com/package/jsonwebtoken) so that unauthorized users are prevented from quering confidential data.

Some additional setup work is needed on your side on the environment variable file under the `backend` directory:
- Set `GCP_CLIENT_ID` to be the client ID from the Google Cloud Platform of your app's OAuth 2.0 Client ID (this client will need to be set up for Web Applications instead of Android Applications)
- Set `GCP_CLIENT_SECRET` to be the client secret from the Google Cloud Platform of your app's OAuth 2.0 Client ID (the same client as above will need to be set up for Web Applications instead of Android Applications)
- Set `JWT_SIGNING_KEY` to be a signing key. If you have NodeJS installed this can be generated using the following on a node shell:`require('crypto').randomBytes(64).toString('hex')`

## How to manually test:
1. Start the server in developer mode via `npm run dev`
2. Through the Postman collection for the server, you should use the request `POST Login Google User` under the `Users` directory. For this request ensure there is no authorization set.
3. The body of the request should look like below. Normally the idToken will be couple with the rest of the information by the frontend when attempting to login. For testing purposes however, until integration is complete you will need to obtain these fields in a hacky way. You will need to slightly change the FE code temporarily by adding `.requestIdToken(GCP_CLIENT_ID)` in the GoogleSignIn options-this should be coming from a Web Application client however not the Android App clients. After doing this, you can log all the following information from when you fetch the account object on the frontend side like shown below.
```commandline
{
    "idToken": <String> - obtain via account.getIdToken();
    "firstName": <String> - obtain via account.getGivenName();
    "lastName": <String> - obtain via account.getGivenName();
    "email": <String> - obtain via account.getEmail();
}
```
4. If you post the request you should either get a status 201 if this user does not currently exist in the database or you will get a status 200 if the user already exists. This completes `Google Sign In Authentication`. Google ID Tokens have an expiry of 1hr so to post logins after a token expires you will need to repeat step 3 again with the update idToken. Normally the frontend will always fetch the idToken in real time so the issue of an expired token should never be a problem.
5. In the above case of a status 201, the frontend response will look like `{userToken: String, user: Object}`. The Frontend will manage the user token on their own, but for our manual testing, we need to keep this token temporarily to authorize future requests in all protected endpoints. Save the token somewhere for now.
6. Go to Postman and for the Vanroomies collection click on `Authorization`. Select type `Bearer` and set the token to the value you saved from before.
7. Now to test all future requests, ensure that in case they use the `authenticateJWT` middleware, you  set `Type` under Authorization to `Inherit from Parent` for that request in Postman - this will utilize the token for that user. A successful request should work just normally whereas if authorization does not wok well you will see the `Unauthorized` error message in lieu of a response.
